Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,287 @@
+        // Preserve the empty map by adding it to the FieldMask.
+        fields.insert(current_path);
+      } else {
+        for (const FieldPath& nested_path : nested_mask) {
+          fields.insert(current_path.Append(nested_path));
+        }
+      }
+    } else {
+      fields.insert(current_path);
+    }
+  }
+  return model::FieldMask(fields);
+}
+
+absl::optional<google_firestore_v1_Value> MutableObjectValue::Get(
+    const firebase::firestore::model::FieldPath& path) const {
+  if (path.empty()) {
+    return *value_;
+  } else {
+    google_firestore_v1_Value nested_value = *value_;
+    for (const std::string& segment : path) {
+      _google_firestore_v1_MapValue_FieldsEntry* entry =
+          FindEntry(nested_value, segment);
+      if (!entry) return {};
+      nested_value = entry->value;
+    }
+    return nested_value;
+  }
+}
+
+void MutableObjectValue::Set(const model::FieldPath& path,
+                             const google_firestore_v1_Value& value) {
+  HARD_ASSERT(!path.empty(), "Cannot set field for empty path on ObjectValue");
+
+  _google_firestore_v1_MapValue* parent_map = ParentMap(path.PopLast());
+
+  std::string last_segment = path.last_segment();
+  std::map<std::string, google_firestore_v1_Value> upserts{
+      {last_segment, value}};
+
+  ApplyChanges(parent_map, upserts, /* deletes= */ {});
+}
+
+void MutableObjectValue::SetAll(const model::FieldMask& field_mask,
+                                const MutableObjectValue& data) {
+  FieldPath parent;
+
+  std::map<std::string, google_firestore_v1_Value> upserts;
+  std::set<std::string> deletes;
+
+  for (const FieldPath& path : field_mask) {
+    if (!parent.IsImmediateParentOf(path)) {
+      // Insert the accumulated changes at this parent location
+      _google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+      ApplyChanges(parent_map, upserts, deletes);
+      upserts.clear();
+      deletes.clear();
+      parent = path.PopLast();
+    }
+
+    absl::optional<google_firestore_v1_Value> value = data.Get(path);
+    if (value) {
+      upserts.emplace(path.last_segment(), *value);
+    } else {
+      deletes.insert(path.last_segment());
+    }
+  }
+
+  _google_firestore_v1_MapValue* parent_map = ParentMap(parent);
+  ApplyChanges(parent_map, upserts, deletes);
+}
+
+google_firestore_v1_MapValue* MutableObjectValue::ParentMap(
+    const FieldPath& path) {
+  google_firestore_v1_Value* parent = value_.get();
+
+  // Find a or create a parent map entry for `path`.
+  for (const std::string& segment : path) {
+    _google_firestore_v1_MapValue_FieldsEntry* entry =
+        FindEntry(*parent, segment);
+
+    if (entry) {
+      if (entry->value.which_value_type !=
+          google_firestore_v1_Value_map_value_tag) {
+        // Since the element is not a map value, free all existing data and
+        // change it to a map type
+        nanopb::FreeNanopbMessage(google_firestore_v1_Value_fields,
```

Comment:
Optional: rather than using `google_firestore_v1_Value_fields` directly, consider using `FieldsArray<google_firestore_v1_Value>()` (from `nanopb/fields_array.h`). It's a little more verbose, but `google_firestore_v1_Value_fields` seems like a really low-level detail that's nice to avoid.